src/sources/kubernetes/file_source_builder.rs

```diff
@@ -0,0 +1,222 @@
+pub struct FileSourceBuilder<'a> {
+    config: &'a KubernetesConfig,
+    file_config: FileConfig,
+}
+
+impl<'a> FileSourceBuilder<'a> {
+    pub fn new(config: &'a KubernetesConfig) -> Self {
+        Self {
+            config,
+            file_config: FileConfig::default(),
+        }
+    }
+
+    pub fn build(
+        mut self,
+        kube_name: &str,
+        globals: &GlobalOptions,
+    ) -> crate::Result<(mpsc::Receiver<Event>, Source)> {
+        self.file_source_include()?;
+        self.file_source_exclude();
+
+        self.file_config.start_at_beginning = true;
+
+        // Filter out files that certainly don't have logs newer than now timestamp.
+        self.file_config.ignore_older = Some(10);
+
+        // oldest_first false, having all pods equaly serviced is of greater importance
+        //                     than having time order guarantee.
+
+        // CRI standard ensures unique naming.
+        self.file_config.fingerprinting = FingerprintingConfig::DevInode;
+
+        // Have a subdirectory for this source to avoid collision of naming its file source.
+        self.file_config.data_dir = Some(globals.resolve_and_make_data_subdir(None, kube_name)?);
+
+        let (file_send, file_recv) = mpsc::channel(1000);
+        let file_source = self
+            .file_config
+            .build("file_source", globals, file_send)
+            .map_err(|e| format!("Failed in creating file source with error: {:?}", e))?;
+
+        Ok((file_recv, file_source))
+    }
+
+    /// Configures include in FileConfig
+    fn file_source_include(&mut self) -> crate::Result<()> {
+        // Prepare include patterns
+
+        // Contains patterns that match on all possible included UIDs.
+        let include_pod_uids = self
+            .numerical_pod_uids()?
+            .into_iter()
+            .flat_map(|uid| to_uid_forms(uid.as_str()))
+            .collect();
+
+        // Pattern that matches to all UIDs, and only UIDs.
+        let any_uid = to_uid_forms("");
+        let pod_uids = not_empty_or_else(&include_pod_uids, &any_uid);
+
+        // Contains patterns: container_name*
+        let include_container_names = self
+            .config
+            .include_container_names
+            .iter()
+            .map(|name| name.clone() + "*")
+            .collect();
+
+        let any_name = vec!["*".to_string()];
+        let container_names = not_empty_or_else(&include_container_names, &any_name);
+
+        let any_namespace = vec!["*".to_string()];
+        // Will match only on exactly included namespaces, or any if none specified.
+        let namespaces = not_empty_or_else(&self.config.include_namespaces, &any_namespace);
+
+        // The following creates and adds path includes of form:
+        // LOG_DIRECTORY/uid/container_name/*.log
+        //
+        // To construct the paths, it's necessary to make every
+        // possible combination of uid and container_name.
+        for uid in pod_uids {
```

Comment:
How does this semantic interact if a user adds a new container to a pod? Since you need both the uid and container to create this path. The idea here is to have them all listed out and hope that itll match? I wonder about things post launch? At that point does it make sense to refresh the container list?

Reply:
If the `include_container_names` filter isn't used, it will match on the new ones, because in that case we are using wildcard `*`. Otherwise only those listed, and in that case users can change the list dynamically with #938.

I hope this answers/clarifies your question.

Reply:
That does, but I also kinda want to think about new uid's that might match filters but wouldn't require a config reload? I think it might make sense to handle that?

Reply:
I am not sure what exactly do you mean. 

If new pod/container starts running on the node, file source will pick it up if it's namespace/uid/name matches to the filters.

Reply:
If a new pod_uid allocates how do we adjust for that?

Reply:
`file_source` will detect and start reading it's log files, nothing else. But that only happens if pod_uid passes include_namespaces and include_pod_uids, else `file_source` will detect it and ignore it.